### PR TITLE
[GSoC'24] Enhacement: Instant Editor UX improvements

### PR DESCRIPTION
**Pull request by criticalAY**
On top of:
*  #16534
## Purpose / Description
- Cloze number to increment only in Increment mode
- Show discard change dialog if there are any changes 
- Share latest text to the note editor 

## How Has This Been Tested?
Tested on Oneplus nord ce

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/instantnoteeditor/InstantNoteEditorActivity.kt [hunk 1]
```diff
@@ -199,9 +204,36 @@ class InstantNoteEditorActivity : AnkiActivity(), DeckSelectionDialog.DeckSelect
                 Timber.d("Save note button pressed")
                 checkAndSave()
             }
-            setOnDismissListener {
-                finish()
+
+            setOnKeyListener { _, keyCode, event ->
+                if (keyCode == KeyEvent.KEYCODE_BACK && event.action == KeyEvent.ACTION_UP) {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/instantnoteeditor/InstantNoteEditorActivity.kt:209]: ⚠️ use `onDismissListener`

File: AnkiDroid/src/main/java/com/ichi2/anki/instantnoteeditor/InstantNoteEditorActivity.kt [hunk 2]
```diff
@@ -199,9 +204,36 @@ class InstantNoteEditorActivity : AnkiActivity(), DeckSelectionDialog.DeckSelect
...
+                    true
+                } else {
+                    false
+                }
+            }
+        }
+
+        // consume the touch event outside the dialog
+        dialogView.rootView.setOnTouchListener { _, event ->
+            if (event.action == MotionEvent.ACTION_DOWN) {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/instantnoteeditor/InstantNoteEditorActivity.kt:224]: invert this condition to make code easier to read

File: AnkiDroid/src/main/java/com/ichi2/anki/instantnoteeditor/InstantNoteEditorActivity.kt [hunk 3]
```diff
@@ -199,9 +204,36 @@ class InstantNoteEditorActivity : AnkiActivity(), DeckSelectionDialog.DeckSelect
...
+                val rect = Rect()
+                dialogView.getGlobalVisibleRect(rect)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/instantnoteeditor/InstantNoteEditorActivity.kt:226]: Consider making a method here to abstract what's happening into a simple concept

```kotlin
if (!dialogView.userClickedOutsideDialog(event)) { return false }
// user clicked inside the dialog - we want to exit
if (intentTextChanged()) {
    showDiscardChangesDialog()
    return@setOnTouchListener true
} else {
    finish()
}
```

File: AnkiDroid/src/main/java/com/ichi2/anki/instantnoteeditor/InstantNoteEditorActivity.kt [hunk 4]
```diff
@@ -199,9 +204,36 @@ class InstantNoteEditorActivity : AnkiActivity(), DeckSelectionDialog.DeckSelect
...
+                    }
+                    true
+                } else {
+                    false
+                }
+            }
+        }
+
+        // consume the touch event outside the dialog
+        dialogView.rootView.setOnTouchListener { _, event ->
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/instantnoteeditor/InstantNoteEditorActivity.kt:223]: ⚠️ Why is this not handled by default?

File: AnkiDroid/src/main/java/com/ichi2/anki/instantnoteeditor/InstantNoteEditorActivity.kt [hunk 5]
```diff
@@ -512,6 +544,17 @@ class InstantNoteEditorActivity : AnkiActivity(), DeckSelectionDialog.DeckSelect
         )
     }
 
+    private fun intentTextChanged(): Boolean {
+        return intent.getStringExtra("extra_text_key") != viewModel.actualClozeFieldText.value
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/instantnoteeditor/InstantNoteEditorActivity.kt:548]: extract `extra_text_key` to a constant so a developer can more easily run _Find Usages_ to understand the code

File: AnkiDroid/src/main/java/com/ichi2/anki/instantnoteeditor/InstantNoteEditorActivity.kt [hunk 6]
```diff
@@ -157,7 +161,7 @@ class InstantNoteEditorActivity : AnkiActivity(), DeckSelectionDialog.DeckSelect
     }
 
     private fun openNoteEditor() {
-        val sharedText = intent.getStringExtra(Intent.EXTRA_TEXT)
+        val sharedText = viewModel.actualClozeFieldText.value
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/instantnoteeditor/InstantNoteEditorActivity.kt:164]: See if you can make a property which exposes `.value`, to reduce overhead reading the code

**Review by david-allison — CHANGES_REQUESTED** [review]
Couple of required comments, rest are implementer's choice, but I'd encourage them

File: AnkiDroid/src/main/java/com/ichi2/anki/instantnoteeditor/InstantNoteEditorActivity.kt (hunk 1 above)
Comment (criticalAY) [reply AnkiDroid/src/main/java/com/ichi2/anki/instantnoteeditor/InstantNoteEditorActivity.kt:209]: The editor dialog would already be dismissed, this is called after the dialog is dismissed so cant use it

File: AnkiDroid/src/main/java/com/ichi2/anki/instantnoteeditor/InstantNoteEditorActivity.kt (hunk 4 above)
Comment (criticalAY) [reply AnkiDroid/src/main/java/com/ichi2/anki/instantnoteeditor/InstantNoteEditorActivity.kt:223]: The touch outside the dialog would dismiss it and if we `setCancellable` to false then the Dialog cant be dismissed by accidental clicks, we will need to handle it on our end

File: AnkiDroid/src/main/java/com/ichi2/anki/instantnoteeditor/InstantNoteEditorActivity.kt (hunk 6 above)
Comment (criticalAY) [reply AnkiDroid/src/main/java/com/ichi2/anki/instantnoteeditor/InstantNoteEditorActivity.kt:164]: get property in activity

**Review by david-allison — APPROVED** [review]
Looks good!

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DiscardChangesDialog.kt [hunk 7]
```diff
@@ -32,6 +33,7 @@ object DiscardChangesDialog {
         message: String = CollectionManager.TR.addingDiscardCurrentInput(),
         positiveMethod: () -> Unit
     ) = AlertDialog.Builder(context).show {
+        Timber.i("showing 'discard changes' dialog")
```
[the remark this answers is not in the document]
Comment (criticalAY) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DiscardChangesDialog.kt:36]: Not temp, this dialog is used in a lot of places so it intentional was put here by David.

**Pull request merged**